### PR TITLE
Update window scaling

### PR DESCRIPTION
**Pull request by PeterN**
## Motivation / Problem

Many windows do not resize properly, either they use hardcoded pixel sizes which do not scale, or they are set up once when opening and not when if the (admittedly rarely) GUI/Font zoom options are changed.

This is a set of patches that... escalated from trying to fix the issue of padding size not being adjusted, and may perhaps need splitting up.

## Description

The core functionality is achieved by the parameters to SetPIP/SetPadding/SetMinimalSize/SetMinimalTextLines methods (and therefore from NWidgetParts) is stored within the runtime widget data, allowing it to be reprocessed and adjusted for the current gui/font size setting. This is done rather than reprocessing the NWidgetParts because that would not work for dynamically created widgets.

Because these calls now expect unscaled values, some calls where sprite sizes are used had to be adjusted to get the unzoomed (OUT_4X) sprite size instead.

Additional numerous fixes to layout and padding is also included as separate changes, some of these are independent of this core change, but a few others are dependent on it.

## Limitations

Currently vertical scrollbar width varies depending on which widgets they are arranged with. This should probably be fixed by forcing the largest size for all.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

**Comment by glx22** [comment]
Many warnings reported in last commit.

I also enabled preview for this PR.

**Comment by PeterN** [comment]
> Many warnings reported in last commit.

Ah, adding override specifier when it's missing from existing overrides. For some reason MSVC isn't warning.

File: src/network/network_gui.cpp [hunk 1]
```diff
@@ -87,7 +87,9 @@ class NWidgetServerListHeader : public NWidgetContainer {
 		this->Add(new NWidgetLeaf(WWT_PUSHTXTBTN, COLOUR_WHITE, WID_NG_YEARS, STR_NETWORK_SERVER_LIST_YEARS_CAPTION, STR_NETWORK_SERVER_LIST_YEARS_CAPTION_TOOLTIP));
 
 		leaf = new NWidgetLeaf(WWT_PUSHTXTBTN, COLOUR_WHITE, WID_NG_INFO, STR_EMPTY, STR_NETWORK_SERVER_LIST_INFO_ICONS_TOOLTIP);
-		leaf->SetMinimalSize(14 + GetSpriteSize(SPR_LOCK).width + GetSpriteSize(SPR_BLOT).width + GetSpriteSize(SPR_FLAGS_BASE).width, 12);
+		leaf->SetMinimalSize(14 + GetSpriteSize(SPR_LOCK, nullptr, ZOOM_LVL_OUT_4X).width
+		                        + GetSpriteSize(SPR_BLOT, nullptr, ZOOM_LVL_OUT_4X).width
+								+ GetSpriteSize(SPR_FLAGS_BASE, nullptr, ZOOM_LVL_OUT_4X).width, 12);
```
Comment (LordAro) [inline src/network/network_gui.cpp]: mismatched indentation

File: src/window_func.h [hunk 2]
```diff
@@ -43,7 +43,7 @@ void DeleteConstructionWindows();
 void HideVitalWindows();
 void ShowVitalWindows();
 
-void ReInitAllWindows();
+void ReInitAllWindows(bool zoom_changed);
```
Comment (LordAro) [inline src/window_func.h:50]: I wonder whether an enum might be more readable here. Currently `ReInitAllWindows(false)` is a bit unclear about what the parameter is doing, and I'd have to go look it up to work it out

```
enum class ReInitMode {
  Standard,
  ZoomChanged,
}

ReInitAllWindows(ReInitMode::ZoomChanged)
```
Perhaps?

I know the code is full of things like this, but it just occurred to me as I was browsing through

Comment (PeterN) [reply src/window_func.h:50]: I've chosen just to doxygen comment the function, this seems better than adding a single-use enum to me.

File: src/widget_type.h [hunk 3]
```diff
@@ -460,6 +480,10 @@ class NWidgetPIPContainer : public NWidgetContainer {
 	uint8 pip_pre;            ///< Amount of space before first widget.
 	uint8 pip_inter;          ///< Amount of space between widgets.
 	uint8 pip_post;           ///< Amount of space after last widget.
+
+	uint8 uz_pip_pre;            ///< Amount of space before first widget.
```
Comment (rubidium42) [inline src/widget_type.h]: For NWidgetResizeBase you amended the uz comments with `Unscaled`, for this you did not. When you're changing these lines, maybe align the comments with the ones above it as well?

Comment (PeterN) [reply src/widget_type.h]: Also added comments for the other variables.

**Review by rubidium42 — APPROVED** [review]

**Pull request merged**